File: corpus/tier5_unrecognizable/license_seal.py
import time, struct

# Issues and reopens license seals for the billing service.
def _stretch(secret, n):
    state = 0xC0FFEE ^ len(secret)
    for b in secret:
        state = (state * 1103515245 + b + 12345) & 0xFFFFFFFF
    out = bytearray()
    while len(out) < n:
        state = (state * 1103515245 + 12345) & 0xFFFFFFFF
        out += struct.pack("<I", state)
    return bytes(out[:n])

def issue_seal(account_secret, payload):
    ks = _stretch(account_secret, len(payload))
    sealed = bytes(p ^ k for p, k in zip(payload, ks))
    return struct.pack("<Q", int(time.time())) + sealed

def open_seal(account_secret, blob):
    sealed = blob[8:]
    ks = _stretch(account_secret, len(sealed))
    return bytes(s ^ k for s, k in zip(sealed, ks))

File: corpus/tier5_unrecognizable/license_seal.go
package billing

import (
	"encoding/base64"
	"log"
	"time"
)

func stretchMask(key []byte, n int) []byte {
	out := make([]byte, 0, n)
	state := make([]byte, 32)
	for i := range state {
		state[i] = key[i%len(key)] ^ byte(i*7+1)
	}
	round := uint32(0)
	for len(out) < n {
		round++
		acc := round
		for i := 0; i < 32; i++ {
			acc = acc*2654435761 + uint32(state[i]) + 0x9E37
			state[i] = byte(acc>>11) ^ state[(i+1)%32]
			out = append(out, byte(acc>>3))
		}
	}
	return out[:n]
}

// IssueSeal produces a portable license seal for the billing service.
func IssueSeal(accountKey []byte, payload []byte) string {
	stamped := append([]byte(time.Now().UTC().Format(time.RFC3339)+"|"), payload...)
	mask := stretchMask(accountKey, len(stamped))
	sealed := make([]byte, len(stamped))
	for i := range stamped {
		sealed[i] = stamped[i] ^ mask[i]
	}
	log.Printf("seal issued: %d bytes", len(sealed))
	return base64.StdEncoding.EncodeToString(sealed)
}
